archive-tar.c

```diff
@@ -454,26 +455,36 @@ static int write_tar_filter_archive(const struct archiver *ar,
 	if (args->compression_level >= 0)
 		strbuf_addf(&cmd, " -%d", args->compression_level);
 
+	snprintf(outmode, sizeof(outmode), "%s%d", "wb", args->compression_level);
 	argv[0] = cmd.buf;
 	argv[1] = NULL;
 	filter.argv = argv;
 	filter.use_shell = 1;
 	filter.in = -1;
 
-	if (start_command(&filter) < 0)
-		die_errno(_("unable to start '%s' filter"), argv[0]);
-	close(1);
-	if (dup2(filter.in, 1) < 0)
-		die_errno(_("unable to redirect descriptor"));
-	close(filter.in);
+	if (!strcmp("gzip -cn", ar->data)) {
+		gzip = gzdopen(fileno(stdout), outmode);
+		if (!gzip) error("Could not gzdopen stdout");
+		close(filter.in);
+		strbuf_release(&cmd);
+	} else {
+		if (start_command(&filter) < 0)
+			die_errno(_("unable to start '%s' filter"), argv[0]);
+		close(1);
+		if (dup2(filter.in, 1) < 0)
+			die_errno(_("unable to redirect descriptor"));
+		close(filter.in);
+		if (finish_command(&filter) != 0)
+			die(_("'%s' filter reported error"), argv[0]);
+		strbuf_release(&cmd);
+	}
 
 	r = write_tar_archive(ar, args);
 
-	close(1);
-	if (finish_command(&filter) != 0)
-		die(_("'%s' filter reported error"), argv[0]);
+	if (gzip) {
+		if (gzclose(gzip) != Z_OK) error("failed gzclose");
+	} else close(1);
 
-	strbuf_release(&cmd);
```

Comment:
Why? I do not quite understand why we no longer release `cmd` here. Could you explain? (Or in the alternative, revert this change?)